decorators_library/decorators.py

```diff
@@ -1 +1,91 @@
+import signal
+import logging
+
 # implement your decorators here.
+class count_calls(object):
+    """
+    Keeps track of how many times certain function was called:
+    """
+    # dictionary to track count of all decorated functions
+    _counters = {}
+    
+    def __init__(self, fn):
+        self._counter = 0
+        self.fn = fn
+        count_calls._counters[self.fn.__name__] = 0
+        
+    def counter(self):
+        return self._counter
+        
+    def __call__(self):
+        self._counter += 1
+        try:
+            self._counters[self.fn.__name__] += 1
+        except KeyError:
+            self._counters[self.fn.__name__] = 1
+        return self.fn()
+
+        
+    @classmethod
+    def counters(cls):
+        return cls._counters
+        
+    @classmethod
+    def reset_counters(cls):
+        cls._counters = {}
+        
+class memoized(object):
+    """
+    This decorator should keep track of previous executions of the decorated 
+    function and the result of the invokations. If the decorated function is 
+    execution again using the same set of arguments sent in the past, 
+    the result must be immediately returned by an internal cache instead 
+    of re executing the same code again. 
+    """
+    def __init__(self, fn):
+        self.cache = {}
+        self.fn = fn
+    
+    def __call__(self, *args):
+        self.cache[args] = self.fn(*args)
+        return self.fn(*args)
+    
```

Comment:
now that I'm looking at your code I'm seeing how much cleaner the code is when implementing classes instead of functions